### PR TITLE
typechecker: A couple of fixes for functions with function parameters

### PR DESCRIPTION
**Pull request by robryanx**

### CODE REVIEW

File: selfhost/typechecker.jakt [hunk 1]
```diff
@@ -2215,6 +2216,51 @@ struct Typechecker {
...
+            Function(params: lhs_params, can_throw: lhs_can_throw, return_type_id: lhs_return_type_id) => {
+                guard rhs_type is Function(params: rhs_params, can_throw: lhs_can_throw, return_type_id: rhs_return_type_id) else {
+                    .error(
+                        format("Type mismatch: expected ‘{}’, but got ‘{}’", .type_name(lhs_type_id), .type_name(rhs_type_id))
+                        span
+                    )
+                    return false
+                }
+
+                if not (lhs_params.size() == rhs_params.size()) {
```
Comment (alimpfard) [inline selfhost/typechecker.jakt:2217]: Let's be nice and say where they differ a bit more precisely, something like "expected function... got function... (the types differ in argument number N, expected i32 got i64)"

File: selfhost/typechecker.jakt [hunk 2]
```diff
@@ -2215,6 +2216,51 @@ struct Typechecker {
                     }
                 }
             }
+            Function(params: lhs_params, can_throw: lhs_can_throw, return_type_id: lhs_return_type_id) => {
```
Comment (alimpfard) [inline selfhost/typechecker.jakt:2208]: Also check if they can throw?

**Review by alimpfard — COMMENTED** [review]
Also need to support it in substitute type vars.

File: selfhost/typechecker.jakt (hunk 1 above)
Comment (alimpfard) [reply selfhost/typechecker.jakt:2217]: Or maybe say that as a hint?

Comment (robryanx) [reply selfhost/typechecker.jakt:2233]: Yeah, added as a hint, it is still using the generic name at the function level as `check_types_for_compat` just returns a bool:
`The parameter types differ at argument 2: expected ‘T’, but got ‘i32’`

There is also the error at the parameter level:
`Type mismatch: expected ‘i64’, but got ‘i32’`

The spans could also probably be improved a bit in future, as there are no spans passed in for the actual parameters.

File: selfhost/typechecker.jakt (hunk 2 above)
Comment (robryanx) [reply selfhost/typechecker.jakt:2207]: Yeah forgot that initially, added now.

**Comment by robryanx** [comment]
> Also need to support it in substitute type vars.

Added, I think it is correct but not sure exactly how I test that case.

**Pull request merged**